app/scripts/lib/createRPCMethodTrackingMiddleware.js

```diff
@@ -0,0 +1,81 @@
+import { SECOND } from '../../../shared/constants/time';
+
+const USER_PROMPTED_EVENT_NAME_MAP = {
+  eth_signTypedData_v4: 'Signature Requested',
+  eth_signTypedData_v3: 'Signature Requested',
+  eth_signTypedData: 'Signature Requested',
+  eth_personal_sign: 'Signature Requested',
+  eth_sign: 'Signature Requested',
+  eth_getEncryptionPublicKey: 'Encryption Public Key Requested',
+  eth_decrypt: 'Decryption Requsted',
+  wallet_requestPermissions: 'Permissions Requested',
+  eth_requestAccounts: 'Permissions Requested',
+};
+
+const samplingTimeouts = {};
+
+/**
+ * Returns a middleware that tracks inpage_provider usage using sampling for
+ * each type of event except those that require user interaction, such as
+ * signature requests
+ *
+ * @param {object} opts - options for the rpc method tracking middleware
+ * @param {Function} opts.trackEvent - trackEvent method from MetaMetricsController
+ * @param {Function} opts.getMetricsState - get the state of MetaMetricsController
+ * @returns {Function}
+ */
+export default function createRPCMethodTrackingMiddleware({
+  trackEvent,
+  getMetricsState,
+}) {
+  return function rpcMethodTrackingMiddleware(
+    /** @type {any} */ req,
+    /** @type {any} */ res,
+    /** @type {Function} */ next,
+  ) {
+    const { origin } = req;
+
```

Comment:
just a suggestion: we could track the request time

```suggestion
const timeStart = Date.now();
```

Reply:
request time like as in the time it takes from when the request comes in to when the response occurs? Or like the actual time the event occurred?

Reply:
the time it takes from when the request comes in to when the response occurs